pycbc/events/stat.py

```diff
@@ -567,6 +568,109 @@ def single(self, trigs):
+        if not len(self.ifos):
+            raise RuntimeError("None of the statistic files has the required "
+                               "attribute called {ifo}-fit_coeffs !")
+        self.fits_by_tid = {}
+        self.alphamax = {}
+        for i in self.ifos:
+            self.fits_by_tid[i] = self.assign_fits(i)
+            self.get_ref_vals(i)
+
+        self.get_newsnr = ranking.get_newsnr
+
+    def assign_fits(self, ifo):
+        coeff_file = self.files[ifo+'-fit_coeffs']
+        template_id = coeff_file['template_id'][:]
+        alphas = coeff_file['fit_coeff'][:]
+        rates = coeff_file['count_in_template'][:]/\
+                float(coeff_file.attrs['analysis_time'])
+        # the template_ids and fit coeffs are stored in an arbitrary order
+        # create new arrays in template_id order for easier recall
+        tid_sort = numpy.argsort(template_id)
+        return {'alpha':alphas[tid_sort], 'rate':rates[tid_sort],
+                'thresh':coeff_file.attrs['stat_threshold']}
+
+    def get_ref_vals(self, ifo):
+        self.alphamax[ifo] = self.fits_by_tid[ifo]['alpha'].max()
+
+    def find_fits(self, trigs):
+        """Get fit coeffs for a specific ifo and template id(s)"""
+        try:
+            tnum = trigs.template_num  # exists if accessed via coinc_findtrigs
+            ifo = trigs.ifo
+        except AttributeError:
+            tnum = trigs['template_id']  # exists for SingleDetTriggers
+            # Should only be one ifo fit file provided
+            assert len(self.ifos) == 1
+            ifo = self.ifos[0]
+        # fits_by_tid is a dictionary of dictionaries of arrays
+        # indexed by ifo / coefficient name / template_id
+        alphai = self.fits_by_tid[ifo]['alpha'][tnum]
+        ratei = self.fits_by_tid[ifo]['rate'][tnum]
+        thresh = self.fits_by_tid[ifo]['thresh']
+        return alphai, ratei, thresh
+
+    def lognoiseratedensity(self, trigs):
+        """
+        Calculate the log noise rate density over single-ifo newsnr
+        Read in single trigger information, make the newsnr statistic
+        and rescale by the fitted coefficients alpha and rate
+        """
+        alphai, ratei, thresh = self.find_fits(trigs)
+        newsnr = self.get_newsnr(trigs)
+        # alphai is constant of proportionality between single-ifo newsnr and
+        #  negative log noise likelihood in given template
+        # ratei is rate of trigs in given template compared to average
+        # thresh is stat threshold used in given ifo
+        lognoisel = numpy.array(numpy.log(alphai) + numpy.log(ratei)
+                                - alphai * (newsnr - thresh),
+                                ndmin=1, dtype=numpy.float32)
+        return lognoisel
+
+    def single(self, trigs):
+        """
+        Single-detector statistic, here just equal to the log noise
+        rate density
+        """
+        return self.lognoiseratedensity(trigs)
```

Comment:
Are these `lognoiseratedensity` and `single` methods different from the existing exp fit stat?